File: framework/appcore/src/MessageFactory.cpp
/* Copyright (c) 2015-2017 CISS, and contributors. All rights reserved.
 *
 * Contributor: Gyeonghwan Hong<redcarrottt@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <libgen.h>

#include "MessageFactory.h"
#include "BaseMessage.h"
#include "cJSON.h"
#include "OPELdbugLog.h"

#define RETURN_IF_INVALID_CJSON_OBJ(a, ret) \
  if((a) == NULL) { \
    OPEL_DBG_ERR("JSON handling error: %s", cJSON_GetErrorPtr()); \
    return ret; \
  }

BaseMessage* MessageFactory::makeMessageFromJSONString(const char* rawString) {
  // Parse rawString in JSON
  cJSON* rootObj = cJSON_Parse(rawString);
  RETURN_IF_INVALID_CJSON_OBJ(rootObj, NULL);
  return makeBaseMessageFromJSON(rootObj);
}

BaseMessage* MessageFactory::makeBaseMessageFromJSON(cJSON* messageObj) {
  cJSON* rootObj = messageObj;

  // Get properties
  // messageId
  cJSON* messageIdObj = cJSON_GetObjectItem(rootObj,
      OPEL_MESSAGE_KEY_MESSAGE_NUM);
  RETURN_IF_INVALID_CJSON_OBJ(messageIdObj, NULL);
  int messageId = atoi(messageIdObj->valuestring);
  
  // URI
  cJSON* uriObj = cJSON_GetObjectItem(rootObj, OPEL_MESSAGE_KEY_URI);
  RETURN_IF_INVALID_CJSON_OBJ(uriObj, NULL);
  std::string uri(uriObj->valuestring);

  // Type
  cJSON* typeObj = cJSON_GetObjectItem(rootObj, OPEL_MESSAGE_KEY_TYPE);
  RETURN_IF_INVALID_CJSON_OBJ(typeObj, NULL);
  // BE AWARE: Since cJSON internally consider integer value as Number(double),
  // I decided to use string-shaped integer in JSON.
  // Once integer value is carried by Number variable, its value can be changed
  // because there is a limit in mantissa's representation.
  BaseMessageType::Value type
    = static_cast<BaseMessageType::Value>(atoi(typeObj->valuestring));

  // isFileAttached
  cJSON* isFileAttachedObj = cJSON_GetObjectItem(rootObj,
      OPEL_MESSAGE_KEY_IS_FILE_ATTACHED);
  RETURN_IF_INVALID_CJSON_OBJ(isFileAttachedObj, NULL);
  int isFileAttachedNum = atoi(isFileAttachedObj->valuestring);
  bool isFileAttached = (isFileAttachedNum == 0) ? false : true;

  // FileName
  std::string fileName("");
  if(isFileAttachedNum) {
    cJSON* fileNameObj = cJSON_GetObjectItem(rootObj,
        OPEL_MESSAGE_KEY_FILE_NAME);
    RETURN_IF_INVALID_CJSON_OBJ(fileNameObj, NULL);
    fileName = fileNameObj->valuestring;
  }

  cJSON* payloadObj = cJSON_GetObjectItem(rootObj, OPEL_MESSAGE_KEY_PAYLOAD);
  RETURN_IF_INVALID_CJSON_OBJ(payloadObj, NULL);

  // Allocate and initialize a new BaseMessage
  BaseMessage* newMessage = NULL;
  switch(type) {
    case BaseMessageType::AppCore:
      newMessage = makeAppCoreMessageFromJSON(payloadObj,
          messageId, uri, isFileAttached, fileName);
      break;
    case BaseMessageType::App:
      newMessage = makeAppMessageFromJSON(payloadObj,
          messageId, uri, isFileAttached, fileName);
      break;
    case BaseMessageType::AppCoreAck:
    case BaseMessageType::Companion:
    default:
      // These types cannot be handled.
      newMessage = new BaseMessage(messageId, uri, type,
          isFileAttached, fileName);
      break;
  }
  return newMessage;
}

AppCoreMessage* MessageFactory::makeAppCoreMessageFromJSON(cJSON* messageObj,
    int messageId, std::string& uri,
    bool isFileAttached, std::string fileName) {
  // Allocate and initialize a new BaseMessage and return it
  cJSON* commandTypeObj = cJSON_GetObjectItem(messageObj,
      APPCORE_MESSAGE_KEY_COMMAND_TYPE);
  RETURN_IF_INVALID_CJSON_OBJ(commandTypeObj, NULL);
  AppCoreMessageCommandType::Value commandType
    = static_cast<AppCoreMessageCommandType::Value>(atoi(
          commandTypeObj->valuestring));

  cJSON* appCorePayloadObj = cJSON_GetObjectItem(messageObj,
      APPCORE_MESSAGE_KEY_PAYLOAD);
  RETURN_IF_INVALID_CJSON_OBJ(appCorePayloadObj, NULL);
  
  // Allocate and initialize a new AppCoreMessage
  AppCoreMessage* newMessage = new AppCoreMessage(
      messageId, uri, isFileAttached, fileName,
      commandType);
  newMessage->setAppCorePayloadObj(appCorePayloadObj);
  return newMessage;
}

AppMessage* MessageFactory::makeAppMessageFromJSON(cJSON* messageObj,
    int messageId, std::string& uri,
    bool isFileAttached, std::string fileName) {
  // Allocate and initialize a new BaseMessage and return it
  cJSON* commandTypeObj = cJSON_GetObjectItem(messageObj, "commandType");
  RETURN_IF_INVALID_CJSON_OBJ(commandTypeObj, NULL);
  AppMessageCommandType::Value commandType
    = static_cast<AppMessageCommandType::Value>(atoi(
          commandTypeObj->valuestring));

  cJSON* appPayloadObj = cJSON_GetObjectItem(messageObj, "payload");
  RETURN_IF_INVALID_CJSON_OBJ(appPayloadObj, NULL);
  
  // Allocate and initialize a new AppMessage
  AppMessage* newMessage = new AppMessage(
      messageId, uri, isFileAttached, fileName, 
      commandType);
  newMessage->setAppPayloadObj(appPayloadObj);
  return newMessage;
}

File: Opel-Manager/src/com/opel/opel_manager/model/message/MessageFactory.java
package com.opel.opel_manager.model.message;

/* Copyright (c) 2015-2017 CISS, and contributors. All rights reserved.
 *
 * Contributor: Gyeonghwan Hong<redcarrottt@gmail.com>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import android.util.Log;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.io.IOException;

public class MessageFactory {
    private static final String TAG = "MessageFactory";

    // string -> JSON -> BaseMessage
    static public BaseMessage makeMessageFromJSONString(String rawString) {
        ObjectMapper mapper = new ObjectMapper();
        JsonNode rootNode = null;
        try {
            rootNode = mapper.readTree(rawString);
        } catch (IOException e) {
            e.printStackTrace();
            Log.e(TAG, "Failed to parse JSON string");
        }
        return makeBaseMessageFromJSON(rootNode);
    }

    // Make AppCoreMessage from bottom
    static AppCoreMessage makeAppCoreMessage(int commandType) {
        return new AppCoreMessage(commandType);
    }

    // JSON -> BaseMessage / AppCoreAckMessage / CompanionMessage
    static BaseMessage makeBaseMessageFromJSON(JsonNode messageObj) {
        ObjectNode thisObj = (ObjectNode) messageObj;

        // messageId
        String messageIdStr = thisObj.get(BaseMessage.OPEL_MESSAGE_KEY_MESSAGE_NUM).asText();
        int messageId = Integer.parseInt(messageIdStr);

        // URI
        String uri = thisObj.get(BaseMessage.OPEL_MESSAGE_KEY_URI).asText();

        // Type
        String typeStr = thisObj.get(BaseMessage.OPEL_MESSAGE_KEY_TYPE).asText();
        int type = Integer.parseInt(typeStr);

        // isFileAttached
        String isFileAttachedStr = thisObj.get(BaseMessage.OPEL_MESSAGE_KEY_IS_FILE_ATTACHED)
                .asText();
        boolean isFileAttached = (Integer.parseInt(isFileAttachedStr) == 1);

        // fileName
        String fileName = thisObj.get(BaseMessage.OPEL_MESSAGE_KEY_FILE_NAME).asText();

        // payload
        JsonNode payloadObj = thisObj.get(BaseMessage.OPEL_MESSAGE_KEY_PAYLOAD);

        // Allocate and initialize a new BaseMessage
        BaseMessage newMessage = new BaseMessage(messageId, uri, type, isFileAttached, fileName);
        switch (type) {
            case BaseMessage.Type_AppCoreAck: {
                AppCoreAckMessage messagePayload = makeAppCoreAckMessageFromJSON(payloadObj);
                newMessage.setPayload(messagePayload);
                break;
            }
            case BaseMessage.Type_Companion: {
                CompanionMessage messagePayload = makeCompanionMessageFromJSON(payloadObj);
                newMessage.setPayload(messagePayload);
                break;
            }
            case BaseMessage.Type_AppCore:
            case BaseMessage.Type_App:
            case BaseMessage.Type_NotDetermined:
            default:
                // These types cannot be handled.
                break;
        }

        return newMessage;
    }

    static AppCoreAckMessage makeAppCoreAckMessageFromJSON(JsonNode messagePayloadObj) {
        ObjectNode thisObj = (ObjectNode) messagePayloadObj;

        // commandMessageId
        String commandMessageIdStr = thisObj.get(AppCoreAckMessage
                .APPCORE_ACK_MESSAGE_KEY_COMMAND_MESSAGE_NUM).asText();
        int commandMessageId = Integer.parseInt(commandMessageIdStr);

        // commandType
        String commandTypeStr = thisObj.get(AppCoreAckMessage
                .APPCORE_ACK_MESSAGE_KEY_COMMAND_TYPE).asText();
        int commandType = Integer.parseInt(commandTypeStr);

        // payload (AppCoreAckMessage's)
        ObjectNode payloadObj = (ObjectNode) thisObj.get(AppCoreAckMessage
                .APPCORE_ACK_MESSAGE_KEY_PAYLOAD);

        // Allocate and initialize a new AppCoreAckMessage
        AppCoreAckMessage newMessage = new AppCoreAckMessage(commandMessageId, commandType);
        newMessage.setPayload(payloadObj);

        return newMessage;
    }

    static CompanionMessage makeCompanionMessageFromJSON(JsonNode messagePayloadObj) {
        ObjectNode thisObj = (ObjectNode) messagePayloadObj;

        // commandType
        String commandTypeStr = thisObj.get(CompanionMessage.COMPANION_MESSAGE_KEY_COMMAND_TYPE)
                .asText();
        int commandType = Integer.parseInt(commandTypeStr);

        // payload (CompanionMessage's)
        ObjectNode payloadObj = (ObjectNode) thisObj.get(CompanionMessage
                .COMPANION_MESSAGE_KEY_PAYLOAD);

        // Allocate and initialize a new AppCoreAckMessage
        CompanionMessage newMessage = new CompanionMessage(commandType);
        newMessage.setPayload(payloadObj);

        return newMessage;
    }
}